### PR TITLE
[WIP] mcu support for stm32g4 

### PR DESCRIPTION
**Pull request by mattthebaker**
Enables stm32g431 mcu support. May not meet 100 user threshold until a general controller board is designed around it. Primary board being addressed initially is the stm32g431 based ([B-G431B-ESC1](https://www.st.com/en/evaluation-tools/b-g431b-esc1.html)) electronic speed controller board for developers that may want to mess with 3 phase motor control.

ADC and UART are tested via ST-Link UART tunnel. SPI and I2C are fairly low risk changes, but board lacks peripherals for testing.

Board cannot support native USB due to how pins are used, but it should work with at most a minor code tweak or two on capable hardware.

I do not have a CAN setup to test, but the board has built in CAN hardware, and based on how the code is structured, should also work with at most a minor code tweak or two. A tester familiar with CAN and having a setup is needed.

Making the work public in case anyone else is wanting to play around with this, there is a foundation to build on.

### CODE REVIEW

**Comment by theolas** [comment]
Thank you for your work!!!

**Comment by KevinOConnor** [comment]
Thanks.  I'm fine with merging this.  (There are likely 100 of the above "esc boards" and the changes appear sufficiently "modular" to me.)

Some minor comments to follow.

-Kevin

File: src/stm32/spi.c [hunk 1]
```diff
@@ -21,14 +21,17 @@ DECL_ENUMERATION("spi_bus", "spi1", 1);
 DECL_CONSTANT_STR("BUS_PINS_spi1", "PA6,PA7,PA5");
 DECL_ENUMERATION("spi_bus", "spi1a", 2);
 DECL_CONSTANT_STR("BUS_PINS_spi1a", "PB4,PB5,PB3");
-#if !CONFIG_MACH_STM32F1
+#if CONFIG_MACH_STM32G4
+ DECL_ENUMERATION("spi_bus", "spi2a", 3);
+ DECL_CONSTANT_STR("BUS_PINS_spi2a", "PA10,PA11,PF1");
+#elif !CONFIG_MACH_STM32F1
```
Comment (KevinOConnor) [inline src/stm32/spi.c:27]: If a new definition is needed, please use the new format - `spi2_PA10_PA11_PF1`.

File: src/stm32/stm32f0_i2c.c [hunk 2]
```diff
@@ -48,6 +48,27 @@ DECL_ENUMERATION("i2c_bus", "i2c2_PB13_PB14", 4);
...
+DECL_ENUMERATION("i2c_bus", "i2c2_PC4_PF0", 5);
+DECL_CONSTANT_STR("BUS_PINS_i2c2_PC4_PF0", "PC4,PF0");
+DECL_ENUMERATION("i2c_bus", "i2c3_PC8_PC9", 6);
+DECL_CONSTANT_STR("BUS_PINS_i2c3_PC8_PC9", "PC8,PC9");
+DECL_ENUMERATION("i2c_bus", "i2c3_PC8_PC11", 7);
+DECL_CONSTANT_STR("BUS_PINS_i2c3_PC8_PC11", "PC8,PC11");
+// requires mixed gpio alternate function indexes
+//DECL_ENUMERATION("i2c_bus", "i2c3_PA8_PB5", 8);
+//DECL_CONSTANT_STR("BUS_PINS_i2c3_PA8_PB5", "PA8,PB5");
+
```
Comment (KevinOConnor) [inline src/stm32/stm32f0_i2c.c:71]: I'm not sure this commented out code is helpful.

File: src/stm32/stm32g4.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
+// Code to setup clocks and gpio on stm32l4
```
Comment (KevinOConnor) [inline src/stm32/stm32g4.c:1]: Stale comment.

File: src/stm32/stm32g4.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    } else /*if (periph_base < AHB)*/ {
+        uint32_t pos = (periph_base - AHB2PERIPH_BASE) / 0x400;
+        return (struct cline){.en = &RCC->AHB2ENR,
+                              .rst = &RCC->AHB2RSTR,
+                              .bit = 1 << pos};
+    }
+  //  return (struct cline){.en = &RCC->AHB2ENR,
+  //                        .rst = &RCC->AHB2RSTR,
+  //                        .bit = 0};
```
Comment (KevinOConnor) [inline src/stm32/stm32g4.c:53]: Dead code.

File: src/stm32/stm32g4.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    // Switch system clock to PLL
+    RCC->CFGR = RCC_CFGR_HPRE_DIV1 | RCC_CFGR_PPRE1_DIV1 | RCC_CFGR_PPRE2_DIV1
+                | RCC_CFGR_SW_PLL;
+    while ((RCC->CFGR & RCC_CFGR_SWS_Msk) != RCC_CFGR_SWS_PLL)
+        ;
+}
+
+// Main entry point - called from armcm_boot.c:ResetHandler()
+void armcm_main(void) {
```
Comment (KevinOConnor) [inline src/stm32/stm32g4.c:155]: Please use consistent indentation.

File: src/stm32/usbfs.c [hunk 6]
```diff
@@ -41,6 +41,11 @@
   #define USB_EPTX_STAT USB_EP_TX_STTX
   #define USB_ISTR_EP_ID USB_ISTR_IDN
   #define USB_CNTR_FRES USB_CNTR_USBRST
+#elif CONFIG_MACH_STM32G4
+  // Transfer memory is accessed with 16bits and contains 16bits of data
+  typedef volatile uint16_t epmword_t;
+  #define WSIZE 2
+  #define USBx_IRQn USB_LP_IRQn
```
Comment (KevinOConnor) [inline src/stm32/usbfs.c:48]: Please move this next to the definition for STM32F0 and STM32L4.

Comment (mattthebaker) [reply src/stm32/usbfs.c]: Turns out G4 supports 32 or 16 bit access to a 16 bit register, so this fits better with the STM32F103 defines due to the IRQ naming.

**Comment by mattthebaker** [comment]
Thanks for the comments and feedback. Quality changes were all made per the suggestions.

Code was rebased onto master to fix merge conflicts in src/stm32/fdcan.c and flash_usb.py

Let me know if the commits should be squashed together or if any other changes should be made before merging.

Thanks,
Matt

**Comment by KevinOConnor** [comment]
Thanks.  Looks fine to me.  Yes, if you could rebase this to squash in the fixes that would be great.  (Ideally keeping the lib/ updates its own separate commit.)

-Kevin

**Comment by KevinOConnor** [comment]
Any further updates on this PR?

-Kevin

**Comment by mattthebaker** [comment]
Merge date estimate sounds good.

Went ahead and switched the files over to the 1.4.0 release. I did review the changes from 1.4.0 to 1.5.0 and they are very low risk, mostly just comments and the license change.

**Pull request merged**